hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/ZKDelegationTokenSecretManager.java

```diff
@@ -520,24 +525,32 @@ protected int incrementDelegationTokenSeqNum() {
     // The secret manager will keep a local range of seq num which won't be
     // seen by peers, so only when the range is exhausted it will ask zk for
     // another range again
-    if (currentSeqNum >= currentMaxSeqNum) {
-      try {
-        // after a successful batch request, we can get the range starting point
-        currentSeqNum = incrSharedCount(delTokSeqCounter, seqNumBatchSize);
-        currentMaxSeqNum = currentSeqNum + seqNumBatchSize;
-        LOG.info("Fetched new range of seq num, from {} to {} ",
-            currentSeqNum+1, currentMaxSeqNum);
-      } catch (InterruptedException e) {
-        // The ExpirationThread is just finishing.. so dont do anything..
-        LOG.debug(
-            "Thread interrupted while performing token counter increment", e);
-        Thread.currentThread().interrupt();
-      } catch (Exception e) {
-        throw new RuntimeException("Could not increment shared counter !!", e);
+    try{
+      this.currentSeqNumLock.lock();
```

Comment:
I am concerned if it is proper here. Because there are some IO operations (request to zk here at L533 and log print atL535) inside the `currentSeqNumLock`, it could decrease performance significantly when meet some network latency or disk busy.
I think it is enough to protect `currentSeqNum` and `currentMaxSeqNum`, right? If that, is it possible to define them as AtomicInteger?

Reply:
Thanks @Hexiaoqiao for your input. 

Here I am trying to address one race condition. This method uses one parameter, seqNumBatchSize. Default is one, with default behaviour it will always hit ZK. But intention of this parameter is to generate the seqNumber in batch to avoid multiple trips to ZK. 

So if the value of this param is set to 500, and just assume it has reached the limit and it's time to regenerate next 500 seqNumber, then it should be not be done by multiple threads. So this regeneration must be made thread safe. That's why, I have acquired this lock.

Now, coming to performance point, if we remove this lock, then, incrSharedCount() method is running in compareAndSet pattern, so chances are very likely that it will try multiple times before updating the value to ZK. So with this lock, all these challenges can be addressed.

If seqNumBatchSize is set to an optimum number, this lock will be acquired just to increment an integer number (most of time ) that should not impact the performance. This will take time only when it interacts with ZK and it will interact with ZK only when already reserved/generated seqNumber gets exhausted.

Please provide your feedback. Thanks.

Reply:
> if we remove this lock, then, incrSharedCount() method is running in compareAndSet pattern, so chances are very likely that it will try multiple times before updating the value to ZK. So with this lock, all these challenges can be addressed.

I don't think it could decrease retry chances where hold lock here, because the znode could be updated by different processes, such as DFSRouter, also if hold lock but retry multiple times, the performance will be impacted tremendously. 

> If seqNumBatchSize is set to an optimum number, this lock will be acquired just to increment an integer number (most of time ) that should not impact the performance. This will take time only when it interacts with ZK and it will interact with ZK only when already reserved/generated seqNumber gets exhausted.

Totally agree, however we give batch size is 1 by default. even through tune this configuration to one bigger number, it could not resolve the above issue.

Thanks @vikaskr22 again. JFYI.

Reply:
> because the znode could be updated by different processes, such as DFSRouter,

is it possible that other processes like DFSRouter can write on these ZNodes ? As far as I know, each process defines their separate znode path, like in KMS we define kms root node path and only KMS instances can write on the specified path. Please correct me if my understanding is wrong.

On race condition, that I mentioned earlier, let me reiterate one more time, consider the following example without any lock:

1. Thread1 has executed the L-532  // currentSeqNum = 10, currentMaxSeqNum=11 
2. Now Thread2 is about to execute the L-531 and it executes it before Thread1 finishes this method. //currentSeqNum = 11, currentMaxSeqNum=12
3. Now Thread1 returns the incremented value, that is 12
4. Then Thread2 returns the incremented value , that is 13. // this is out of range.
In this case, the returned value would be out of range. I mean, it would be greater than currentMaxSeqNum, that would create functional issues. because same sequenceNumber might be reserved by other instances.

I hope I am able to clarify the scenario

@Hexiaoqiao , please confirm if this race condition is possible or not? If Yes, then following solution could be considered:

1. Make currentMaxSeqNum, currentSeqNumLock AtomicInteger // as suggested by you. 
2. Now, instead of simple lock,, use ReentrantReadWriteLock currentSeqNumLock;
3.   Now in incrementDelegationTokenSeqNum() method, following pattern could be used:
      
     currentSeqNumLock.readLock().lock();
    if (currentSeqNum.get() >= currentMaxSeqNum.get()) {
             currentSeqNumLock.readLock().unlock();  // release the read lock
            currentSeqNumLock.writeLock().lock();
            // Double check the condition one more time:
            if (currentSeqNum.get() >= currentMaxSeqNum.get()) {
                // Now increment the seqNum on ZK and update the variables.
           }
       }
    finally unlock the acquired locks.
  
here, all threads will be able to acquire readLock and will not block each other. AtomicInteger will ensure atomicity and visibility issue. And write lock will address the race condition and it will be required only when it needs to interact with ZK.

Above is a simple pseudo code, please review this approach and share your feedback.

And thanks for your review efforts.

Reply:
Hi @Hexiaoqiao , Can you please provide your input on this ? I have further updated the code.

Reply:
@vikaskr22 Thanks for your works and sorry for the late response. After carefully review. I think it is possible of the race condition you mentioned above. And before changes, there is one `synchronized` to protect this critical section.
Back to this PR, I prefer to the first version, because the performance could not improve obviously but with lock escalations. What do you think about?
Let's wait if other guys would like to have some other opinions. cc @jojochuang

Reply:
@Hexiaoqiao , Actually in earlier code, most of the API level methods like verifyToken, createToken etc were marked synchronized and that was blocking the threads that are not even related.

This particular method was not protected by  "synchronized"  keyword, it's super-class method has synchronization defined. But we were not getting any race situation because the method from where it is being called was marked synchronized.

And yes, I agree that earlier approach where we had taken simple ReentrantLock at start of method execution was a simple & clean approach. I can revert back to that. 

Thanks.

Reply:
Yeah.... a simple ReentrantLock looks like the way to go.

But also

>  however we give batch size is 1 by default.
oh.... i regret this. (zk-dt-secret-manager.token.seqnum.batch.size) is a configuration key not documented anywhere. How would folks discover this. 

@vikaskr22 did you need to update this configuration key in order to achieve the performance improvement?

Reply:
@jojochuang , Yes. We did benchmarking for KMS and there We got at least 3X better throughput with `zk-dt-secret-manager.token.seqnum.batch.size=1000` .

I discovered this while going through the code.